include/seqan3/io/alignment_file/input.hpp

```diff
@@ -0,0 +1,1200 @@
+     *
+     * ### Example
+     *
+     * \snippet test/snippet/io/alignment_file/alignment_file_input.cpp sam_file
+     *
+     * \snippet test/snippet/io/alignment_file/alignment_file_input.cpp get_header
+     *
+     * \sa seqan3::alignment_file_header
+     */
+    alignment_file_header & header()
+    {
+        return *header_ptr;
+    }
+
+protected:
+    //!\privatesection
+
+    //!\brief The file header object.
+    std::unique_ptr<alignment_file_header> header_ptr{new alignment_file_header()};
+
+    /*!\name Data buffers
+     * \{
+     */
+    //!\brief Buffer for a single record.
+    record_type record_buffer;
+    //!\}
+
+    /*!\name Stream / file access
+     * \{
+     */
+    //!\brief The type of the internal stream pointers. Allows dynamically setting ownership management.
+    using stream_ptr_t = std::unique_ptr<std::basic_istream<stream_char_type>,
+                                         std::function<void(std::basic_istream<stream_char_type>*)>>;
+    //!\brief Stream deleter that does nothing (no ownership assumed).
+    static void stream_deleter_noop(std::basic_istream<stream_char_type> *) {}
+    //!\brief Stream deleter with default behaviour (ownership assumed).
+    static void stream_deleter_default(std::basic_istream<stream_char_type> * ptr) { delete ptr; }
+
+    //!\brief The primary stream is the user provided stream or the file stream if constructed from filename.
+    stream_ptr_t primary_stream{nullptr, stream_deleter_noop};
+    //!\brief The secondary stream is a compression layer on the primary or just points to the primary (no compression).
+    stream_ptr_t secondary_stream{nullptr, stream_deleter_noop};
+
+    //!\brief File is at position 1 behind the last record.
+    bool at_end{false};
+
+    //!\brief Type of the format, an std::variant over the `valid_formats`.
+    using format_type = detail::transfer_template_args_onto_t<valid_formats, std::variant>;
+
+    //!\brief The actual std::variant holding a pointer to the detected/selected format.
+    format_type format;
+    //!\}
+
+    //!\brief The reference ids corresponding to the alignment records.
+    std::unordered_map<ref_id_type, size_t> ref_ids_to_position;
+
+    //!\brief The reference sequences corresponding to the alignment records.
+    std::vector<ref_sequence_type> const * reference_sequences_ptr{nullptr};
+
+    /*!\brief Set reference information (only important for SAM/BAM reading).
+     *
+     * \details
+     *
+     * The SAM format only provides semi-alignments because the reference sequence
+     * is not stored explicitly. In order to be able to read in full alignment,
+     * additional reference information can be given to the alignment file.
+     * Note that the reference ids (names) must correspond to the exact spelling
+     * in the SAM/BAM file otherwise an exception will be thrown when reading.
+     *
+     * ### Example
+     *
+     * \todo example
```

Comment:
👀

Reply:
It's in details so I will not provide an example for time reasons